rfcs/20210129-optimize-pd-creation/README.md

```diff
@@ -0,0 +1,262 @@
+3. Increases memory consumption
+4. New API to control one more cache
+
+The proposal is to implement an internal mechanism that would significantly
+reduce time of primitive descriptor creation but do not expose the mechanism to user.
+
+The idea is to use existing cache for primitives to implement the mechanism
+because the cache already contains all the abstractions that are created during those
+3 steps. The `primitive_t` abstraction contains `primitive_desc_t` which contains
+`op_desc_t` and `primitive_attr_t`.
+According to the primitive cache design there was added an ability to share
+the `primitive_desc_t` abstraction with `shared_ptr` therefore we can simply use it
+to create a new `primitive_desc_t` with minimum overhead.
+
+Since primitive descriptor and primitive are created for each execution of an
+operation in frameworks and the approach to caching primitives works it will
+also work for primitive descriptors.
+
+## Design
+
+Basically, the cache related work is done as part of enabling the primitive cache.
+The main goal now is to provide a bridge between API that is responsible for creating
+primitive descriptors and the cache for primitives.
+
+### Key for Primitive Descriptor
+
+The key which is used to find primitives in the cache can be used to find primitive
+descriptors as well however, there are a few problems to solve.
+
+Currently, the key looks as follows:
+```cpp
+struct key_t {
+     key_t(const primitive_desc_t *pd, const engine_t *engine, int impl_nthr);
+    // Obtained from `primitive_desc_t`
+    op_desc_t *op_desc_;
+    primitive_attr_t *attr_;
+    std::type_index impl_id_;
+    std::vector<memory_desc_t> bwd_mds;
+
+    // Obtained from `op_desc_t`
+    primitive_kind_t primitive_kind_;
+
+    // Obtained from `engine_t`
+    engine_kind_t engine_kind_;
+    runtime_kind_t runtime_kind_;
+    device_id_t device_id_;
+
+    // Obtained from ctor
+    int impl_nthr_;
+
+    // ... //
+```
+
+When creating a key to find a primitive in the cache there is a primitive descriptor
+which is used to obtain some of the required data. However, in case of primitive descriptor
+creation for all primitives except reorder, concat and sum user creates `op_desc_t`,
+`primitive_attr_t`, `engine_t` and optionally a hint primitive descriptor. Therefore
+`impl_id_`, `bwd_mds` and `op_desc_t` for the aforementioned primitives cannot
+be obtained at that point.
+
+#### Implementation ID
+
+The rationale behind having an implementation ID is to be able to differentiate
+the same primitives which have different underlying implementations. Such primitives
+can be created when user uses primitive descriptor iterator to manually iterate
+through the list of implementations.
+
+The problem here is that the primitive descriptor is only going to be created and
+currently there is no way to get implementation ID without creating primitive
+descriptor.
+
+Currently, `std::type_index` of a primitive descriptor of a particular implementation
```

Comment:
Are you referring to `primitive_desc_t`?

I think it may also be worth introducing a small section (somewhere at the beginning of this RFC), where you specify how to distinguish between user-facing entities (dnnl::primitive{_desc}) and non-user-facing ones (primitive{_desc}_t). Otherwise I find it challenging to distinguish the two from the context.

In case you refer only to a single entity through this document, please specify it explicitly.

Reply:
No, I'm not referring to `primitive_desc_t`. The idea is to state class name explicitly when it's needed otherwise use the abstraction name.

This sentence is referring to `jit_avx512_common_convolution_fwd_t::pd_t`, `jit_avx2_convolution_fwd_t::pd_t`, etc. Since it's not about a particular class  the abstraction name is used.

Typically, the assumption for architecture related RFCs is that the reader is familiar with oneDNN architecture. Sometimes it's worth it to describe some additional details but in this particular case I don't think the description you are suggesting is appropriate because  it's not important in this context.